core/trino-main/src/main/java/io/trino/sql/planner/LocalExecutionPlanner.java

```diff
@@ -3475,12 +3476,19 @@ private int getWriterCount(Session session, Optional<PartitioningScheme> partiti
                 return 1;
             }
 
-            // The default value of partitioned writer count is 32 which is high enough to use it
-            // for both cases when scaling is enabled or not. Additionally, it doesn't lead to too many
-            // small files since when scaling is disabled only single writer will handle a single partition.
-            return partitioningScheme
-                    .map(scheme -> getTaskPartitionedWriterCount(session))
-                    .orElseGet(() -> isLocalScaledWriterExchange(source) ? getTaskScaleWritersMaxWriterCount(session) : getTaskWriterCount(session));
+            WriterScalingOptions scalingOptions = writerTarget.getWriterScalingOptions(metadata, session);
+            if (partitioningScheme.isPresent()) {
+                // The default value of partitioned writer count is 32 which is high enough to use it
+                // for both cases when scaling is enabled or not. Additionally, it doesn't lead to too many
+                // small files since when scaling is disabled only single writer will handle a single partition.
+                return isLocalScaledWriterExchange(source)
+                        ? scalingOptions.taskScalingMaxWriterCount().orElse(getTaskPartitionedWriterCount(session))
+                        : getTaskPartitionedWriterCount(session);
+            }
+
+            return isLocalScaledWriterExchange(source)
+                    ? scalingOptions.taskScalingMaxWriterCount().orElse(getTaskScaleWritersMaxWriterCount(session))
```

Comment:
I think you should take `min` from scaling options and feature session (here and above)